frame/pallet-xvm/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+        input: Vec<u8>,
+        metadata: Vec<u8>,
+    ) -> Result<Vec<u8>, Vec<u8>> {
+        for_tuples!( #(
+            if Tuple::id() == context.id {
+                log::trace!(
+                    target: "xvm::SyncVm::xvm_call",
+                    "VM found, run XVM call: {:?}, {:?}, {:?}, {:?}, {:?}",
+                    context, from, to, input, metadata,
+                );
+                return Tuple::xvm_call(context, from, to, input, metadata)
+            }
+        )* );
+        log::trace!(
+            target: "xvm::SyncVm::xvm_call",
+            "VM with ID {:?} not found", context.id
+        );
+        Err(b"VM is not found".to_vec())
+    }
+}
+
+/// The engine that support asynchronous smart contract execution.
+/// For example, XCVM.
+pub trait AsyncVM<VmId, AccountId> {
+    /// Unique VM identifier.
+    fn id() -> VmId;
+
+    /// Send a message.
+    fn xvm_send(
+        context: XvmContext<VmId>,
+        from: AccountId,
+        to: Vec<u8>,
+        message: Vec<u8>,
+        metadata: Vec<u8>,
+    ) -> bool;
+
+    /// Query for incoming messages.
+    fn xvm_query(context: XvmContext<VmId>, inbox: AccountId) -> Vec<Vec<u8>>;
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(30)]
+impl<VmId: Member + Default, AccountId: Member> AsyncVM<VmId, AccountId> for Tuple {
+    fn id() -> VmId {
+        Default::default()
+    }
+
+    fn xvm_send(
+        context: XvmContext<VmId>,
+        from: AccountId,
+        to: Vec<u8>,
+        message: Vec<u8>,
+        metadata: Vec<u8>,
+    ) -> bool {
+        for_tuples!( #(
+            if Tuple::id() == context.id {
+                log::trace!(
+                    target: "xvm::AsyncVM::xvm_send",
+                    "VM found, send message: {:?}, {:?}, {:?}, {:?}, {:?}",
+                    context, from, to, message, metadata,
+                );
+                return Tuple::xvm_send(context, from, to, message, metadata)
+            }
+        )* );
+        log::trace!(
+            target: "xvm::AsyncVM::xvm_send",
+            "VM with ID {:?} not found", context.id
+        );
+        false
+    }
+
+    fn xvm_query(context: XvmContext<VmId>, inbox: AccountId) -> Vec<Vec<u8>> {
+        for_tuples!( #(
+            if Tuple::id() == context.id {
+                log::trace!(
+                    target: "xvm::AsyncVM::xvm_query",
+                    "VM found, query messages: {:?} {:?}",
+                    context, inbox,
+                );
+                return Tuple::xvm_query(context, inbox)
+            }
+        )* );
+        log::trace!(
+            target: "xvm::AsyncVM::xvm_query",
+            "VM with ID {:?} not found", context.id
+        );
+        Default::default()
+    }
+}
+
+/// Universal VM parameter codec.
+///
+/// Input data is SCALE encoded, output is specified for each VM.
+pub trait XvmCodec {
+    /// Convert generic XVM input into VM native.
+    fn encode(input: Vec<u8>, metadata: Vec<u8>) -> Vec<u8>;
```

Comment:
I believe this should be `decode`?

`encode` should return encoded data and metadata, right?

Reply:
Encode to native, decode from generic, so, it’s same meaning. May be better use `convert` here.